src/swarm_algorithms/whale_algorithm.py

```diff
@@ -0,0 +1,66 @@
+from src.swarm_algorithms.swarm_algorithm import SwarmIntelligence
+
+import numpy as np
+import scipy.spatial.distance as dist
+
+
+class WhaleAlgorithm(SwarmIntelligence):
+    def __init__(self, population_size, nb_features, constraints, attenuation_of_medium,
+                 intensity_at_source, seed=0xCAFFE):
+        super().__init__(population_size, nb_features, constraints, seed)
+        self._eta = attenuation_of_medium
+        self._rho_zero = intensity_at_source
+
+    def get_new_positions(self, step_number):
+        nearest_best_whales, distances_to_best_whales = self._get_nearest_best_whale_and_distance(self.population)
+        return self.population + self._rng.uniform(0, self._rho_zero * np.exp(
+            -self._eta * distances_to_best_whales[:, np.newaxis]), size=(self.population_size, self.nb_features)) * (
+                       self.population[nearest_best_whales] - self.population)
+
+    def update_positions(self, new_positions, step_number):
+        self.population = new_positions
+        return self.population
+
+    def go_swarm_go(self):
+        # TODO: Maybe implement in the future
+        pass
+
+    def _get_nearest_best_whale_and_distance(self, population):
+        """
+        Calculates nearest best whale for every whale in the population.
+        :param population: Population to consider.
+        :return: Indices of best whales in the population for each whale as np.ndarray N shape
+        """
+        assert self.is_compiled
+        fit_values = self.fit_function(population)
+        nearest_best_whales = []
+        distances_to_best_whales = []
+        for i in range(len(fit_values)):
+            where_better = np.where(fit_values >= fit_values[i][0])[0]
+            considered_whales = population[where_better]
+            distances = dist.cdist(considered_whales, population[i:i + 1])[1:, 0]
+            if len(distances) == 0:
+                nearest_best_whales.append(i)
+                distances_to_best_whales.append(0)
+                continue
+            index = np.argmax(distances)
+            distances_to_best_whales.append(distances[index])
+            nearest_best_whales.append(index)
+        nearest_best_whales = np.array(nearest_best_whales)
+        distances_to_best_whales = np.array(distances_to_best_whales)
+        return nearest_best_whales, distances_to_best_whales
+
+    @staticmethod
+    def get_optimal_eta_and_rho_zero(search_space_boundaries):
```

Comment:
Where will it be used? I assume that in `go_swarm_go`, if so where will boundaries come from? 
as a parameter to `go_swarm_go`? If so - is it applicant to all swarm algorithms or just this particular one?